zebra-script/src/lib.rs

```diff
@@ -2,10 +2,155 @@
 #![doc(html_logo_url = "https://www.zfnd.org/images/zebra-icon.png")]
 #![doc(html_root_url = "https://doc.zebra.zfnd.org/zebra_script")]
 
+use displaydoc::Display;
+use thiserror::Error;
+use zcashconsensus::{
+    zcashconsensus_error_t, zcashconsensus_error_t_zcashconsensus_ERR_OK,
+    zcashconsensus_error_t_zcashconsensus_ERR_TX_DESERIALIZE,
+    zcashconsensus_error_t_zcashconsensus_ERR_TX_INDEX,
+    zcashconsensus_error_t_zcashconsensus_ERR_TX_SIZE_MISMATCH,
+};
+use zebra_chain::{
+    block::Block,
+    transaction::{TransparentInput, TransparentOutput},
+};
+
+#[derive(Debug, Display, Error)]
+#[non_exhaustive]
+pub enum Error {
+    /// Verification failed but tx was valid
+    #[non_exhaustive]
+    ErrOk,
+    /// Could not to deserialize tx
+    #[non_exhaustive]
+    TxDeserialize,
+    /// n_in is invalid for tx
+    #[non_exhaustive]
+    TxIndex,
+    /// tx is an invalid size for it's protocol
+    #[non_exhaustive]
+    TxSizeMismatch,
+    /// Encountered unknown error kind from zcashconsensus: {0}
+    #[non_exhaustive]
+    Unknown(zcashconsensus_error_t),
+}
+
+impl From<zcashconsensus_error_t> for Error {
+    #[allow(non_upper_case_globals)]
+    fn from(err_code: zcashconsensus_error_t) -> Error {
+        match err_code {
+            zcashconsensus_error_t_zcashconsensus_ERR_OK => Error::ErrOk,
+            zcashconsensus_error_t_zcashconsensus_ERR_TX_DESERIALIZE => Error::TxDeserialize,
+            zcashconsensus_error_t_zcashconsensus_ERR_TX_INDEX => Error::TxIndex,
+            zcashconsensus_error_t_zcashconsensus_ERR_TX_SIZE_MISMATCH => Error::TxSizeMismatch,
+            unknown => Error::Unknown(unknown),
+        }
+    }
+}
+
+pub fn verify_script(
+    script_pub_key: impl AsRef<[u8]>,
+    amount: i64,
+    tx_to: impl AsRef<[u8]>,
+    n_in: u32,
+    flags: u32,
+    consensus_branch_id: u32,
+) -> Result<(), Error> {
```

Comment:
Can we workshop the signature of this function a little bit? I think we should aim to expose a function whose signature is oriented around the `zebra-chain` types, and does whatever translation is necessary internally.

Reply:
Some questions on the types in this method:

- should `script_pub_key` be a `&Script` instead of `impl AsRef<[u8]>`?
- should `amount` be an `Amount` instead of an `i64`?  If so, what constraint? My guess is that we can accept a generic constraint here, but we should check.
- what is `tx_to`?
- what is `flags`?
- how can we eliminate `consensus_branch_id`? Elsewhere in zebra we don't have a type for this, treating it as a serialization detail, and we should continue that in this API.

Reply:
Can we merge this function with the one below, and present only a nicely-typed method?  I think that we will need to nail down the interface here before we can be sure about the async script design in #868 .

Reply:
oh yea, this one isn't supposed to be pub, this is just the ffi safety wrapper that mirrors the API of the original function. I'd prefer to keep it separate from the other fn but we should definitely not make it pub.